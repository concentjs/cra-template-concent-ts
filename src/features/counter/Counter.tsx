import React, { useState } from 'react';
import { useC2Mod } from 'configs/useC2Mod';
import { COUNTER } from 'configs/c2Mods';
import styles from './Counter.module.css';

export function Counter() {
  const ctx = useC2Mod(COUNTER);
  // module state, module computed, module reducer
  const { state, moduleComputed, mr } = ctx;
  const [incrementAmount, setIncrementAmount] = useState('2');

  return (
    <div>
      <div className={styles.row}>
        <button
          className={styles.button}
          aria-label="Increment value"
          onClick={mr.increment}
        >
          +
        </button>
        <span className={styles.value}>{state.value}</span>
        <span className={styles.value}>doubleCount: {moduleComputed.doubleCount}</span>
        <button
          className={styles.button}
          aria-label="Decrement value"
          onClick={mr.decrement}
        >
          -
        </button>
      </div>
      <div className={styles.row}>
        <input
          className={styles.textbox}
          aria-label="Set increment amount"
          value={incrementAmount}
          onChange={e => setIncrementAmount(e.target.value)}
        />
        <button
          className={styles.button}
          onClick={() => mr.incrementByAmount(Number(incrementAmount) || 0)}
        >
          Add Amount
        </button>
        <button
          className={styles.asyncButton}
          onClick={() => mr.incrementAsync(Number(incrementAmount) || 0)}
        >
          Add Async
        </button>
        <button
          className={styles.normalButton}
          onClick={mr.incrementBigValue}
        >
          Add Big Value
        </button>
      </div>
    </div>
  );
}
