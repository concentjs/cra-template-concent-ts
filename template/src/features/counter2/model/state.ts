
function getInitialState() {
  return {
    value: 6,
    bigValue: 100,
    toInc: 2,
  };
}

export type St = ReturnType<typeof getInitialState>;

export default getInitialState;
