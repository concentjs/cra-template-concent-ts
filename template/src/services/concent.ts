/**
 * concent 相关的一些公共封装函数
 */
import {
  useConcent, reducer,
  getState as getSt, getGlobalState as getGst, emit,
  ReducerCallerParams, IReducerFn, IActionCtxBase,
  ICtxBase, IAnyObj, SettingsType, SetupFn,
} from 'concent';
import { CtxM, Modules, RootRd, RootState, CtxDe } from 'types/store';
import { EvMap } from 'types/eventMap';

/**
 * 调用目标函数，用于对接 reducer里的 ghost函数
 * @param callerParams 
 * @param ac 
 */
export async function callTarget(callerParams: ReducerCallerParams | [IReducerFn, any], ac: IActionCtxBase) {
  try {
    // 支持 reducer文件里内部调用 ac.dispatch(loading, [targetFn, payload])
    if (Array.isArray(callerParams)) {
      let [fn, payload] = callerParams;
      await ac.dispatch(fn, payload);
    } else {
      const { fnName, payload, renderKey, delay } = callerParams;
      await ac.dispatch(fnName, payload, renderKey, delay);
    }
  } catch (err) {
    alert(err.message);
  }
}

/**
 * 适用于不属于任何模块，只是设置setup函数的场景
 * @param setup 
 */
export function useSetup<T extends SetupFn>(setup: T, props?: any) {
  const { settings } = useConcent<{}, CtxDe<{}, SettingsType<T>>>({ setup, props });
  return settings;
}

/**
 * use the target model context you want by passing a module name
 * 如需要全局任意地方可通过 useC2Mod('xx') 导出xx模块上下文来使用，
 * 需要在 src/models/index.js 显式的导出该模块
 * 
 * -----------------------[Code example]-----------------------
 *  import somePageModel from 'pages/SomePage/model';
 *  import someCompModel from 'components/SomeComp/model';
 * 
 *  const allModels = {
 *    ...somePageModel,
 *    ...someCompModel,
 *  }
 * 
 *  export default allModels;
 * --------------------------------------------------------------
 * @param moduleName
 */
export function useC2Mod<M extends Modules>(moduleName: M) {
  type Ctx = CtxM<{}, M>;
  return useConcent<{}, Ctx>(moduleName);
}


export function useC2ModWithSetup<M extends Modules, Setup extends (ctx: ICtxBase) => IAnyObj, P extends IAnyObj>
  (module: M, setup: Setup, props?: P) {
  const targetProps = props || {};
  type Ctx = CtxM<P, M, SettingsType<Setup>>;
  return useConcent<{}, Ctx>({ module, setup, props: targetProps });
}

export const ccReducer = (reducer as unknown) as RootRd;

/**
 * 获取 globa模块的状态
 * 在已拥有concent model上下文、action上线文的地方，推荐直接获取，代替调用此函数
 */
export function getGlobalState() {
  const globalState = getGst<RootState>();
  return globalState;
}

/**
 * 获取整个根状态
 */
export function getRootState() {
  const rootState = getSt() as RootState;
  return rootState;
}

/**
 * 获取目标模块状态
 */
export function getModelState<T extends Modules>(modelName: T) {
  const modelState = getSt(modelName) as RootState[T];
  return modelState;
}

type EvKeys = keyof EvMap;
/**
 * 发射事件
 * @param eventName - 事件名
 * @param args 
 */
export function ccEmit<E extends EvKeys, T extends EvMap[E]>(eventName: E, ...args: T) {
  emit(eventName, ...args);
}

/**
 * 携带id的发射事件
 * @param eventdesc - [eventName, id]
 * @param args 
 */
export function ccEmitId<E extends EvKeys, T extends EvMap[E]>(eventdesc: [E, string], ...args: T) {
  emit(eventdesc, ...args);
}

type OnFn = <E extends EvKeys>(eventName: E, cb: (...args: EvMap[E]) => void) => void;
/**
 * 配合EvMap，为ctx.on装配类型
 * 外部调用时传入具体的事件名就推导出cb函数的参数列表类型
 * 
 *  function setup(ctx: Ctx){
 *    const on = ctxOn(ctx);
 *    on('xxx',(a, b)=>{
 *      // 此处ts能感知a、b的具体类型
 *    })
 *  }
 * 
 */
export function ctxOn(ctx: ICtxBase) {
  return ctx.on as OnFn;
}

type OnIdFn = <E extends EvKeys>(eventdesc: [E, string], cb: (...args: EvMap[E]) => void) => void;
/**
 * 可以携带id的ctx.on
 * @param ctx 
 */
export function ctxOnId(ctx: ICtxBase) {
  return ctx.on as OnIdFn;
}
