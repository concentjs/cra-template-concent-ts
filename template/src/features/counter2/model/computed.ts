import { St } from './state';

// only value change will triiger this function to execute again
export function doubleCount({ value }: St) {
  return value * 2;
}
