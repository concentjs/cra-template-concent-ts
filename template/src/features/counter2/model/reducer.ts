import { St } from './state';
import { IAC } from './meta';
import { VoidPayload } from 'types/store';

const delay = (ms = 1000) => new Promise(r => setTimeout(r, ms));

export function incrementBigValue(payload: VoidPayload, moduleState: St): Partial<St> {
  return { bigValue: moduleState.bigValue + 50 };
}

export function incrementOne(payload: VoidPayload, moduleState: St): Partial<St> {
  return { value: moduleState.value + 1 };
}

export function decrementOne(payload: VoidPayload, moduleState: St): Partial<St> {
  return { value: moduleState.value - 1 };
}

export function increment(payload: VoidPayload, moduleState: St): Partial<St> {
  return { value: moduleState.value + moduleState.toInc };
}

export async function incrementAsync(payload: VoidPayload, moduleState: St, ac: IAC): Promise<Partial<St>> {
  await delay();
  // or just write ac.dispatch of return
  // await ac.dispatch(incrementByAmount, amount);
  return { value: moduleState.value + moduleState.toInc };
}

export function foo() {
  console.log('call foo');
}

export function clear() {
  console.log('call clear');
}
