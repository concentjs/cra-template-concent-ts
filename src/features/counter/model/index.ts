import { COUNTER } from '../../../configs/c2Mods';
import state from './state';
import * as computed from './computed';
import * as reducer from './reducer';

export default {
  [COUNTER]: {
    state,
    computed,
    reducer,
  }
}
