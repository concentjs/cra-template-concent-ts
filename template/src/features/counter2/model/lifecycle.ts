import { IDispatch } from 'concent';
import * as reducer from './reducer';


// 异步状态初始化函数
// export async function initState(){}


// 当前模块的第一个组件实例挂载完毕时触发
// 默认值触发一次，如需满足条件反复触发，return false即可
export function mounted(dispatch: IDispatch) {
  dispatch(reducer.foo);
}

// 当前模块的最后一个组件实例卸载时触发
// 默认值触发一次，如需满足条件反复触发，return false即可
export function willUnmount(dispatch: IDispatch) {
  dispatch(reducer.clear);
}