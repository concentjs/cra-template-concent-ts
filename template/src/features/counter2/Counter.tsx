import React from "react";
import { useModel } from './model/meta';
import styles from "./Counter.module.css";

export function Counter() {
  const ctx = useModel();
  // module state, module computed, module reducer
  const { state, moduleComputed, mr } = ctx;

  return (
    <div>
      <div className={styles.row}>
        <button
          className={styles.button}
          aria-label="Increment value"
          onClick={mr.increment}
        >
          +
        </button>
        <span className={styles.value}>{state.value}</span>
        <span className={styles.value}>
          doubleCount: {moduleComputed.doubleCount}
        </span>
        <button
          className={styles.button}
          aria-label="Decrement value"
          onClick={mr.decrementOne}
        >
          -
        </button>
      </div>
      <div className={styles.row}>
        <input
          className={styles.textbox}
          aria-label="Set increment amount"
          value={state.toInc}
          onChange={(e) => mr.setState({ toInc: Number(e.target.value) || 0 })}
        />
        <button
          className={styles.button}
          onClick={mr.increment}
        >
          Add Amount For Counter2
        </button>
        <button
          className={styles.asyncButton}
          onClick={mr.incrementAsync}
        >
          Add Async
        </button>
        <button className={styles.normalButton} onClick={mr.incrementBigValue}>
          Add Big Value
        </button>
      </div>
    </div>
  );
}
