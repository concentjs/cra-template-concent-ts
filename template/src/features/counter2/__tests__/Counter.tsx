import React from 'react';
import { mount, ReactWrapper } from 'enzyme';
import { dispatch } from 'concent';
import * as rd from '../model/reducer';
import { St } from '../model/meta';
import { Counter } from '../Counter';

// 组件实例
let ins = null as unknown as ReactWrapper;

/**
 * @author fantasticsoul
 */
describe('Counter Render', () => {
  const getInsValue = () => {
    const valueWrap = ins.find('span').at(0);
    return parseInt(valueWrap.text(), 10);
  };

  const testIncrementBtnAction = (expectedValue: number) => {
    const incBtnWrap = ins.find('button').at(0);
    incBtnWrap.simulate('click');
    const actualValue = getInsValue();
    expect(actualValue === expectedValue).toBeTruthy();
  };

  beforeAll(() => {
    ins = mount(<Counter />);
  });

  test('generate AllocateGroup snapshot', () => {
    expect(ins).toMatchSnapshot();
  });

  test('should have 5 buttons', () => {
    const btnWrap = ins.find('button');
    expect(btnWrap.length === 5).toBeTruthy();
  });

  test('mr.increment should work', async () => {
    // const { value } = await ccReducer.Counter.increment(null, { silent: true });
    // or
    const { value } = await dispatch<St>(rd.increment, null, { silent: true });
    testIncrementBtnAction(value);
  });

});