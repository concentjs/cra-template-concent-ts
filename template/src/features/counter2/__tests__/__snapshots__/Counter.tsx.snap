// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Render generate AllocateGroup snapshot 1`] = `
<Counter>
  <div>
    <div
      className="row"
    >
      <button
        aria-label="Increment value"
        className="button"
        onClick={[Function]}
      >
        +
      </button>
      <span
        className="value"
      >
        0
      </span>
      <span
        className="value"
      >
        doubleCount: 
        0
      </span>
      <button
        aria-label="Decrement value"
        className="button"
        onClick={[Function]}
      >
        -
      </button>
    </div>
    <div
      className="row"
    >
      <input
        aria-label="Set increment amount"
        className="textbox"
        onChange={[Function]}
        value={2}
      />
      <button
        className="button"
        onClick={[Function]}
      >
        Add Amount
      </button>
      <button
        className="asyncButton"
        onClick={[Function]}
      >
        Add Async
      </button>
      <button
        className="normalButton"
        onClick={[Function]}
      >
        Add Big Value
      </button>
    </div>
  </div>
</Counter>
`;
